src/cls/rbd/cls_rbd.cc

```diff
@@ -2664,13 +2669,13 @@ int metadata_list(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
     for (; it != raw_data.end(); ++it)
       data[metadata_name_from_key(it->first)].swap(it->second);
 
-    if (r < max_read)
+    if (!more)
       break;
 
     last_read = raw_data.rbegin()->first;
     if (max_return)
       max_read = MIN(RBD_MAX_KEYS_READ, max_return - data.size());
-  } while (max_read);
+  } while (more);
```

Comment:
Note: since ```more``` is set to true when ```max_return``` is reached, this effectively breaks the functionality of ```max_return``` and instead returns all matching omap values. I opened tracker ticket 21247 to audit the remainder of the rbd cls methods to ensure ```max_return``` is being correctly honored.